plugins/modules/packaging/os/redhat_subscription.py

```diff
@@ -593,15 +593,22 @@ def update_subscriptions_by_pool_ids(self, pool_ids):
         consumed_pools = RhsmPools(self.module, consumed=True)
 
         existing_pools = {}
+        serials_to_remove = []
         for p in consumed_pools:
-            existing_pools[p.get_pool_id()] = p.QuantityUsed
+            pool_id = p.get_pool_id()
+            quantity_used = p.get_quantity_used()
+            existing_pools[pool_id] = quantity_used
+
+            quantity = pool_ids.get(pool_id, 0)
+            if quantity == 0 or quantity is not None and quantity != quantity_used:
```

Comment:
Can you explain this condition (and the one for the `if` below you've changed)? I don't think I understand what you are trying to achieve here.

Reply:
Of course. I have to admit, that this part has caused some headaches on the way to find a working solution. This and the condition below are vital to make the module idempotent.

The newly introduced method `get_quantity_used()` makes sure that the quantity that was previously returned by calling the subscription-manager via CLI is an integer and no longer a string. Comparing a string with an integer will not give us the result, we are expecting. In the next step, we assign the current quantity to the pool to which the system is already subscribed to, so that `existing_pools` contains all subscribed pools with the corresponding quantities. `pool_ids` contains all pools to which the system should be subscribed to, combined with a quantity represented as an integer or `None`.

Now comes the tricky part, because we must compare the required state with the current state and find out, if we must do something or not. `pool_ids.get(pool_id, 0)` returns an integer or `None`. `0` means that the system should now longer be subscribed to the corresponding pool, e.g. when the user made a reconfiguration after all entitlements got successfully established in a previous run. `None` means colloquially that the subscription-manager should decide about the exact amount of entitlements required to subscribe the system. When we get `None` here, we know that the system is subscribed to the correct pool, but we cannot decide if the amount of entitlements is correct. In this case, we assume the correctness of the already established entitlements. A problem only occurs, when the user increased the socket number after the system was subscribed and when no quantities are provided. At the very end, we compare the current quantity (int) with the required quantity (int) and only when they are unequal we know that we must increase or decrease the number of active entitlements.

Reply:
The integer conversion ensures the idempotency of the module when all quantities are integers. Unfortunately, the idempotency problem in case of `None` was not addressed by the commit https://github.com/ansible/ansible/pull/66933/files where the default quantity was changed from `1` to `None`, when the user does not provide any.

Reply:
Hmm, but then why is the first part of the condition needed? `quantity is not None and quantity != quantity_used` should be totally sufficient, because if `quantity == 0` implies `quantity is not None`, and since `quantity_used > 0` (my assumption, not sure whether it is true) we also have `quantity != quantity_used`. But even if `quantity_used == 0`, if `quanity == 0` we probably don't want to remove the serial since its quantity is already `0`?